### ISSUE TITLE
[feature request] Any way to control the order of send/recv (host to device data transfer) explicitly?

### ISSUE THREAD

**Issue by jiazhe0909** [tensorflow/tensorflow#12818]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
source
- **TensorFlow version (use command below)**:
1.2
- **Python version**:
3.5
- **Bazel version (if compiling from source)**:
0.5.2
- **CUDA/cuDNN version**:
8.0
- **GPU model and memory**:
GTX 1070

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

Overlapping the host to device data transfer with GPU calculation is an important method to improve the performance of inference workload. But I found it very difficult to control the order of data transfers explicitly at this moment. I understand the send/recv operations are added implicitly when the original graph is split/optimized into sub-graphs. I tried to use control dependency + identity/assign operation (as suggested [here](https://github.com/tensorflow/tensorflow/issues/2848)) to hand control send/recv order. But it seems impossible to achieve this goal. 

For example, 
In the following code, we'd like to overlap the H2D memcpy of B->B_GPU with the matmul calculation.
```
import os
import tensorflow as tf
import numpy as np

from tensorflow.python.client import session
from tensorflow.python.framework import ops
from tensorflow.python.ops import variables

os.environ['TF_CPP_MIN_LOG_LEVEL'] = '1'

with ops.Graph().as_default():
  with tf.device('/cpu:0'):
    A = tf.placeholder(tf.float32, shape=(1000, 100), name="A")
    B = tf.placeholder(tf.float32, shape=(1000, 1000), name="B")
  with tf.device('/gpu:0'):
    C = variables.Variable(tf.random_normal([100,1000],stddev=0.1), name="C")
    A_GPU = tf.identity(A)
    with tf.control_dependencies([A_GPU]):
      B_GPU = tf.identity(B)
      D = tf.add(tf.matmul(A_GPU, C), B_GPU, name="D")

  random_A=np.random.rand(1000,100).astype(np.float32)
  random_B=np.random.rand(1000,1000).astype(np.float32)

  sess = session.Session()
  init = tf.global_variables_initializer()
  sess.run(init)

  for x in range(1,50):
    output = sess.run(D,feed_dict={A:random_A,B:random_B})
```

But it turned out that H2D memcpy A->A_GPU is not ensured to launch first. Actually among the 50 iterations, I noticed the order of transferring A->A_GPU and transferring B->B_GPU is randomly performed if we use multiple threads (because the two transfers are handled in different threads, and no-dependency could be built between them). 
If A->A_GPU launches first, the matmul op could be overlapped with B->B-GPU
![image](https://user-images.githubusercontent.com/8175586/30090954-3f28ef90-92a6-11e7-98ee-2a05e6a7463d.png)
otherwise, the matmul op has to wait until all H2D memcpys finished.
![image](https://user-images.githubusercontent.com/8175586/30091033-c6b2f988-92a6-11e7-8abb-f41feac3bbf1.png)

Only if we can control the order of send node explicitly, the overlap can be ensured.

Please let me know if I didn't use the identity operation correctly.

Thanks.

**Comment by cy89** [comment]
@poxvoculi can you offer any hints for this use case?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by poxvoculi** [comment]
First, I apologize for not noticing this issue earlier.  In case it's still of interest, here's some analysis.

Short summary: It's complicated. I'll give a longer answer than strictly necessary, to address cases more general than you describe.

Control edges introduce a dependency where if there is such an edge from node i to node j, node j will not be scheduled for execution until node i has 'finished' execution.  I put the scare quotes around finished because while it's true in the way you probably expect for CPU operations, it's not really true in the case of GPU operations.  

CPU operations are scheduled by immediately executing them or by queuing them on a threadpool, and their completion is detected synchronously or via callback.  

GPU operations are scheduled by making a request to the StreamExecutor (an abstract wrapper around the CUDA driver).  TF pre-establishes a small number of execution streams on each GPU, for dedicated purposes.  All compute kernels are executed on a single compute stream.  Since that is the case, if nodes i and j in the example are both to execute on the same GPU, we can schedule j immediately after scheduling i, without waiting for i to finish, because j won't actually begin execution until i completes.   (In general we don't know or care exactly when a stream operation completes, just that the proper stream event ordering dependencies are introduced to guarantee safe execution.)  H2D copy operations are scheduled on their own dedicated stream, D2H copy operations on another, and D2D copy operations on yet another.  So it's possible to have up to 4 concurrent operations involving one GPU: one compute and three copies.  At least from the TF point of view; at the NVIDIA device level there may be a more detailed description with a different character.

When a TF graph is executed, it is first partitioned into a subgraph of only nodes assigned to a single device, and value copy operations between those graphs are introduced in the form of (internal only) Send and Recv nodes.  Unless there's an optional control input a Recv node is eligible for immediate execution, but the data transfer doesn't start until it meets the corresponding Send in the Rendezvous table associated with the process of the source device.

With these considerations, let's examine your program.  After the first run C is fully initialized and the only operations that execute in each run are 
   1. feeding A  
   2. feeding B
   3. copying A to A_GPU
   4. copying B to B_GPU
   5. doing the matmul
   6. doing the add
Only operations 1 & 2 execute on the CPU, all the rest are GPU stream operations.  I don't really know how the feeding works; if it's multi-threaded I suppose it may be non-deterministic whether A or B is completely populated first.

Your program adds control edges from the tf.identity that establishes A_GPU to the tf.identity that establishes B_GPU and also the matmul and the add.  Strictly speaking, the first tf.identity executes on the GPU so it's eligible to schedule as soon as the Recv of A is 'finished'.  The second tf.identity is also a GPU operation and is eligible to schedule as soon as the Recv of B is finished *and* the first tf.identity is 'finished'.  Note one tricky point however; the data copy is triggered by the corresponding Send and Recv meeting in the process Rendezvous table, and your control edge is between two tf.identity operations, both of which are actually downstream (in terms of the dataflow graph) from the Recv operations.  From the point-of-view of the GPU executor, both of those Recv nodes are eligible for immediate scheduling, and in your multi-threaded environment it appears to be non-deterministic which copy gets into the H2D stream first.  Once it's in there, it will run to completion before the other copy starts.

The semantics of tf.identity are potentially confusing where the input tensor is non-local and the purpose of the operation is to introduce a copy.  In general, identity creates a new immutable tensor which shares type, shape and the same backing memory buffer with its input tensor.  Where the given input is non-local, the execution environment introduces an upstream Recv which actually allocates the memory buffer and populates it, and that Recv op is technically the input to the identity op.   If this is happening on a GPU then the Recv-triggered copy is scheduled on the H2D stream and the identity is scheduled on the compute stream. To ensure that the identity doesn't execute too soon, we could insert a dependency on the H2D stream in the compute stream immediately before enqueuing the identity op.  However, instead we treat Recv specially as the only async GPU operation and actually wait for it to terminate before considering it to be finished and enqueuing the identity op.

In the case of your program, the main problem is that you've introduced the control dependencies downstream of the actual data copies, but it's tricky to realize your intended behavior because of the strange semantics of GPU operations.  I would try rewriting your program as follows:

      with tf.device('/gpu:0'):
        A_GPU = tf.identity(A)
      with tf.device('/cpu:0'):
        with tf.control_dependencies([A_GPU]):
          B_staged = tf.identity(B)
      with tf.device('gpu:0'):
        M = tf.matmul(A_GPU, C)
        B_GPU = tf.identity(B_staged)
        D = tf.add(M, B_GPU, name="D")

I find that with this definition the A copy consistently happens before the B copy, and the matmul may overlap with the B copy, BUT: overall execution is slower.  That's because starting the B copy stalls on waiting for a stream event to be harvested out of the CUDA driver and handled by the CPU, and in this particular program the matmul is relatively fast compared to copy times.  Also, maybe the B populate finishes first a lot of the time, but now we stall waiting for the B populate to finish before doing anything. 
 But, maybe you can now find a way to improve your real problem.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by poxvoculi** [comment]
closing for lack of activity

**Comment by WencongXiao** [comment]
@poxvoculi  You suggestion is great and it works! I have one more question related to the same topic. Do you have any idea about how to control the dependency of ops in the auto-grad(backpropagation)?
